trunk/paradiseo-eo/doc/man/man3/eoParser.3
.TH "eoParser" 3 "19 Oct 2006" "Version 0.9.4-cvs" "EO" \" -*- nroff -*-
.ad l
.nh
.SH NAME
eoParser \- eoParser: command line parser and configuration file reader This class is persistent, so it can be stored and reloaded to restore parameter settings.  

.PP
.SH SYNOPSIS
.br
.PP
\fC#include <eoParser.h>\fP
.PP
Inherits \fBeoParameterLoader\fP, \fBeoObject\fP, and \fBeoPersistent\fP.
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBeoParser\fP (unsigned _argc, char **_argv, std::string _programDescription='', std::string _lFileParamName='param-file', char _shortHand= 'p')"
.br
.RI "\fIConstructor a complete constructor that reads the command line an optionally reads a configuration file. \fP"
.ti -1c
.RI "void \fBprocessParam\fP (\fBeoParam\fP &param, std::string section='')"
.br
.RI "\fIProcesses the parameter and puts it in the appropriate section for readability. \fP"
.ti -1c
.RI "void \fBreadFrom\fP (std::istream &is)"
.br
.RI "\fIRead object. \fP"
.ti -1c
.RI "void \fBprintOn\fP (std::ostream &os) const "
.br
.RI "\fIWrite object. \fP"
.ti -1c
.RI "std::string \fBclassName\fP (void) const "
.br
.RI "\fIclassName for readibility \fP"
.ti -1c
.RI "bool \fBuserNeedsHelp\fP (void)"
.br
.RI "\fItrue if the user made an error or asked for help \fP"
.ti -1c
.RI "void \fBprintHelp\fP (std::ostream &os)"
.br
.RI "\fIPrints an automatic help in the specified output using the information provided by parameters. \fP"
.ti -1c
.RI "std::string \fBProgramName\fP ()"
.br
.ti -1c
.RI "virtual bool \fBisItThere\fP (\fBeoParam\fP &_param) const "
.br
.RI "\fIHas param been entered by user? \fP"
.ti -1c
.RI "\fBeoParam\fP * \fBgetParamWithLongName\fP (const std::string &_name) const "
.br
.RI "\fIget a handle on a param from its longName \fP"
.ti -1c
.RI "template<class ValueType> \fBeoValueParam\fP< ValueType > & \fBgetORcreateParam\fP (ValueType _defaultValue, std::string _longName, std::string _description, char _shortHand=0, std::string _section='', bool _required=false)"
.br
.RI "\fIGet or create parameter. \fP"
.ti -1c
.RI "template<class ValueType> \fBeoValueParam\fP< ValueType > & \fBsetORcreateParam\fP (ValueType _defaultValue, std::string _longName, std::string _description, char _shortHand=0, std::string _section='', bool _required=false)"
.br
.RI "\fISet parameter value or create parameter. \fP"
.ti -1c
.RI "void \fBsetStopOnUnknownParam\fP (bool _b)"
.br
.RI "\fIaccessors to the stopOnUnknownParam value \fP"
.ti -1c
.RI "bool \fBgetStopOnUnknownParam\fP ()"
.br
.ti -1c
.RI "void \fBsetPrefix\fP (const std::string &_prefix)"
.br
.RI "\fIPrefix handling. \fP"
.ti -1c
.RI "void \fBresetPrefix\fP ()"
.br
.ti -1c
.RI "std::string \fBgetPrefix\fP ()"
.br
.in -1c
.SS "Private Types"

.in +1c
.ti -1c
.RI "typedef std::multimap< std::string, \fBeoParam\fP * > \fBMultiMapType\fP"
.br
.ti -1c
.RI "typedef std::map< char, std::string > \fBShortNameMapType\fP"
.br
.ti -1c
.RI "typedef std::map< std::string, std::string > \fBLongNameMapType\fP"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBdoRegisterParam\fP (\fBeoParam\fP &param) const "
.br
.ti -1c
.RI "std::pair< bool, std::string > \fBgetValue\fP (\fBeoParam\fP &_param) const "
.br
.ti -1c
.RI "void \fBupdateParameters\fP () const "
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "MultiMapType \fBparams\fP"
.br
.ti -1c
.RI "std::string \fBprogramName\fP"
.br
.ti -1c
.RI "std::string \fBprogramDescription\fP"
.br
.ti -1c
.RI "ShortNameMapType \fBshortNameMap\fP"
.br
.ti -1c
.RI "LongNameMapType \fBlongNameMap\fP"
.br
.ti -1c
.RI "\fBeoValueParam\fP< bool > \fBneedHelp\fP"
.br
.ti -1c
.RI "\fBeoValueParam\fP< bool > \fBstopOnUnknownParam\fP"
.br
.ti -1c
.RI "std::vector< std::string > \fBmessages\fP"
.br
.ti -1c
.RI "std::string \fBprefix\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
eoParser: command line parser and configuration file reader This class is persistent, so it can be stored and reloaded to restore parameter settings. 
.PP
Definition at line 100 of file eoParser.h.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "eoParser::eoParser (unsigned _argc, char ** _argv, std::string _programDescription = \fC''\fP, std::string _lFileParamName = \fC'param-file'\fP, char _shortHand = \fC'p'\fP)"
.PP
Constructor a complete constructor that reads the command line an optionally reads a configuration file. 
.PP
myEo --param-file=param.rc will then load using the parameter file param.rc
.PP
\fBParameters:\fP
.RS 4
\fI_argc\fP command line arguments count 
.br
\fI_argv\fP command line parameters 
.br
\fI_programDescription\fP Description of the work the program does 
.br
\fI_lFileParamName\fP Name of the parameter specifying the configuration file (--param-file) 
.br
\fI_shortHand\fP Single charachter shorthand for specifying the configuration file 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "void eoParser::readFrom (std::istream & is)\fC [virtual]\fP"
.PP
Read object. 
.PP
\fBParameters:\fP
.RS 4
\fI_is\fP A std::istream. 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIruntime_std::exception\fP If a valid object can't be read. 
.RE
.PP

.PP
Implements \fBeoPersistent\fP.
.SS "void eoParser::printOn (std::ostream & os) const\fC [virtual]\fP"
.PP
Write object. 
.PP
It's called printOn since it prints the object on a stream. 
.PP
\fBParameters:\fP
.RS 4
\fI_os\fP A std::ostream. 
.RE
.PP

.PP
Implements \fBeoPrintable\fP.
.SS "virtual bool eoParser::isItThere (\fBeoParam\fP & _param) const\fC [inline, virtual]\fP"
.PP
Has param been entered by user? 
.PP
Checks if _param has been actually entered by the user 
.PP
Implements \fBeoParameterLoader\fP.
.PP
Definition at line 148 of file eoParser.h.
.PP
Referenced by setORcreateParam().
.SS "\fBeoParam\fP* eoParser::getParamWithLongName (const std::string & _name) const"
.PP
get a handle on a param from its longName 
.PP
if not found, returns 0 (null pointer :-)
.PP
Not very clean (requires hard-coding of the long name twice!) but very useful in many occasions... 
.PP
Referenced by getORcreateParam().
.SS "template<class ValueType> \fBeoValueParam\fP<ValueType>& eoParser::getORcreateParam (ValueType _defaultValue, std::string _longName, std::string _description, char _shortHand = \fC0\fP, std::string _section = \fC''\fP, bool _required = \fCfalse\fP)\fC [inline]\fP"
.PP
Get or create parameter. 
.PP
It seems finally that the easiest use of the above method is through the following, whose interface is similar to that of the widely-used createParam.
.PP
For some (probably very stupid) reason, I failed to put it in the .cpp. Any hint??? 
.PP
Definition at line 173 of file eoParser.h.
.PP
References eoParameterLoader::createParam(), and getParamWithLongName().
.SS "template<class ValueType> \fBeoValueParam\fP<ValueType>& eoParser::setORcreateParam (ValueType _defaultValue, std::string _longName, std::string _description, char _shortHand = \fC0\fP, std::string _section = \fC''\fP, bool _required = \fCfalse\fP)\fC [inline]\fP"
.PP
Set parameter value or create parameter. 
.PP
This makes sure that the specified parameter has the given value. If the parameter does not exist yet, it is created.
.PP
This requires that operator<< is defined for ValueType.
.PP
\fBParameters:\fP
.RS 4
\fI_defaultValue\fP Default value. 
.br
\fI_longName\fP Long name of the argument. 
.br
\fI_description\fP Description of the parameter. 
.br
\fI_shortName\fP Short name of the argument (Optional) 
.br
\fI_section\fP Name of the section where the parameter belongs. 
.br
\fI_required\fP Is the parameter mandatory? 
.RE
.PP
\fBReturns:\fP
.RS 4
Corresponding parameter. 
.RE
.PP

.PP
Definition at line 211 of file eoParser.h.
.PP
References eoParameterLoader::createParam(), isItThere(), and eoValueParam< ValueType >::setValue().

.SH "Author"
.PP 
Generated automatically by Doxygen for EO from the source code.


eo/NEWS
* release 1.0 (not yet released)
  - Implement CMA-ES.
  - Update introductory pages of documentation and webpage.
  - Add Microsoft Visual C++ project support files (for Visual Studio 2003 and 2005)
  - Upgrade Teamplates/ script to create a complete standalon EO project (using autotools)
  - Remove support for pre-standard C++ compiler (i.e. gcc-2.x), which allows to
    clean up the code considerably. 
    Assume availability of sstream and limits.

* release 0.9.3z.1 (1. Oct. 2005)
  - Support gcc-3.4 and gcc.4.x.
  - Provide full automake/autoconf/configure support.


# Local Variables:
# coding: iso-8859-1
# mode: text
# fill-column: 80
# End:


trunk/paradiseo-eo/src/utils/eoParser.cpp
// -*- mode: c++; c-indent-level: 4; c++-member-init-indent: 8; comment-column: 35; -*-

//-----------------------------------------------------------------------------
// eoParser.cpp
// (c) Marc Schoenauer, Maarten Keijzer and GeNeura Team, 2000
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

    Contact: todos@geneura.ugr.es, http://geneura.ugr.es
             Marc.Schoenauer@inria.fr
             mkeijzer@dhi.dk
 */
//-----------------------------------------------------------------------------

#ifdef _MSC_VER
#pragma warning(disable:4786)
#endif

#include <stdexcept>
#include <algorithm>
#include <fstream>
#include <iomanip>

#include <utils/compatibility.h>

#include <utils/eoParser.h>

using namespace std;

void eoWarning(std::string str)
{
    cout << str << '\n';
}

std::ostream& printSectionHeader(std::ostream& os, std::string section)
{
    if (section == "")
        section = "General";

    os << '\n' << setw(10) << "######    " << setw(20) << section << setw(10) << "    ######\n";
    return os;
}

eoParameterLoader::~eoParameterLoader()
{
    for (unsigned i = 0; i < ownedParams.size(); ++i)
    {
        delete ownedParams[i];
    }
}


eoParser::eoParser ( unsigned _argc, char **_argv , string _programDescription,
                     string _lFileParamName, char _shortHand) :
    programName(_argv[0]),
    programDescription( _programDescription),
    needHelp(false, "help", "Prints this message", 'h'),
    stopOnUnknownParam(true, "stopOnUnknownParam", "Stop if unkown param entered", '\0')
{
    // need to process the param file first
    // if we want command-line to have highest priority
    unsigned i;
    for (i = 1; i < _argc; ++i)
    {
	if(_argv[i][0] == '@')
        { // read response file
	    char *pts = _argv[i]+1; // yes a char*, sorry :-)
	    ifstream ifs (pts);
	    ifs.peek(); // check if it exists
	    if (!ifs)
            {
		string msg = string("Could not open response file: ") + pts;
		throw runtime_error(msg);
            }
	    // read  - will be overwritten by command-line
	    readFrom(ifs);
	    break; // stop reading command line args for '@'
        }
    }
    // now read arguments on command-line
    stringstream stream;
    for (i = 1; i < _argc; ++i)
    {
        stream << _argv[i] << '\n';
    }
    readFrom(stream);
    processParam(needHelp);
    processParam(stopOnUnknownParam);
}



eoParam * eoParser::getParamWithLongName(const std::string& _name) const
{
    typedef std::multimap<std::string, eoParam*> MultiMapType;
    typedef MultiMapType::const_iterator iter;
    std::string search(prefix+_name);
    for(iter p = params.begin(); p != params.end(); ++p)
        if(p->second->longName() == search)
            return p->second;
    return 0;
}



void eoParser::processParam(eoParam& param, std::string section)
{
    // this param enters the parser: add the prefix to the long name
    if (prefix != "")
    {
	param.setLongName(prefix+param.longName());
	section = prefix + section;  // and to section
    }
    doRegisterParam(param); // plainly register it
    params.insert(make_pair(section, &param));
}

void eoParser::doRegisterParam(eoParam& param) const
{
    if (param.required() && !isItThere(param))
    {
        string msg = "Required parameter: " + param.longName() + " missing";
        messages.push_back(msg);
    }
    pair<bool, string> value = getValue(param);
    if (value.first)
    {
        param.setValue(value.second);
    }
}

pair<bool, string> eoParser::getValue(eoParam& _param) const
{
    pair<bool, string> result(false, "");

    if (_param.shortName() != 0)
    {
        map<char, string>::const_iterator it = shortNameMap.find(_param.shortName());
        if (it != shortNameMap.end())
        {
            result.second = it->second;
            result.first = true;
            return result;
        }
    }
    map<string, string>::const_iterator it = longNameMap.find(_param.longName());
    if (it != longNameMap.end())
    {
        result.second = it->second;
        result.first = true;
        return result;
    }
    // else (TODO: check environment, just long names)
    return result;
}

void eoParser::updateParameters() const
{
 typedef MultiMapType::const_iterator It;

  for (It p = params.begin(); p != params.end(); ++p)
  {
        doRegisterParam(*p->second);
  }
}

void eoParser::readFrom(istream& is)
{
    string str;
    // we must avoid processing \section{xxx} if xxx is NOT "Parser"
    bool processing = true;
    while (is >> str)
    {
        if (str.find(string("\\section{"))==0) // found section begin
            processing = (str.find(string("Parser"))<str.size());

        if (processing)		// right \section (or no \section at all)
	{
            if (str[0] == '#')
	    { // skip the rest of the line
                string tempStr;
                getline(is, tempStr);
	    }
            if (str[0] == '-')
	    {
                if (str.size() < 2)
		{
                    eoWarning("Missing parameter");
                    needHelp.value() = true;
                    return;
		}

                if (str[1] == '-') // two consecutive dashes
		{
                    string::iterator equalLocation = find(str.begin() + 2, str.end(), '=');
                    string value;

                    if (equalLocation == str.end())
		    { // TODO: it should be the next string
                        value = "";
		    }
                    else
		    {
                        value = string(equalLocation + 1, str.end());
		    }

                    string name(str.begin() + 2, equalLocation);
                    longNameMap[name] = value;

		}
                else // it should be a char
		{
                    string value = "1"; // flags do not need a special

                    if (str.size() >= 2)
		    {
                        if (str[2] == '=')
			{
                            if (str.size() >= 3)
                                value = string(str.begin() + 3, str.end());
			}
                        else
			{
                            value = string(str.begin() + 2, str.end());
			}
		    }

                    shortNameMap[str[1]] = value;
		}
	    }
	}
    }

    updateParameters();
}

void eoParser::printOn(ostream& os) const
{
    typedef MultiMapType::const_iterator It;

    It p = params.begin();

    std::string section = p->first;

    printSectionHeader(os, section);
    //print every param with its value
    for (; p != params.end(); ++p)
      {
        std::string newSection = p->first;

        if (newSection != section)
        {
            section = newSection;
            printSectionHeader(os, section);
        }

        eoParam* param = p->second;

	if (!isItThere(*param))  // comment out the ones not set by the user
	  os << "# ";

        string str = "--" + param->longName() + "=" + param->getValue();

        os.setf(ios_base::left, ios_base::adjustfield);
        os << setw(40) << str;

        os << setw(0) << " # ";
        if (param->shortName())
            os << '-' << param->shortName() << " : ";
        os << param->description();

        if (param->required())
        {
            os << " REQUIRED ";
        }

        os  << '\n';
    }
}

void eoParser::printHelp(ostream& os)
{
    if (needHelp.value() == false && !messages.empty())
    {
        std::copy(messages.begin(), messages.end(), ostream_iterator<string>(os, "\n"));
        messages.clear();
        return;
    }

    // print program name and description
    os << this->programName <<": "<< programDescription << "\n\n";

    // print the usage when calling the program from the command line
    os << "Usage: "<< programName<<" [Options]\n";
    // only short usage!
    os << "Options of the form \"-f[=Value]\" or \"--Name[=value]\"" << endl;

    os << "Where:"<<endl;

    typedef MultiMapType::const_iterator It;

    It p = params.begin();

    std::string section = p->first;

    printSectionHeader(os, section);

    //print every param with its value
    for (; p != params.end(); ++p)
    {
        std::string newSection = p->first;

        if (newSection != section)
        {
            section = newSection;
            printSectionHeader(os, section);
        }

        if (p->second->shortName())
	        os << "-" << p->second->shortName() << ", ";

        os << "--" <<p->second->longName() <<":\t"
	     << p->second->description() ;

	  os << "\n" << setw(20) << ( (p->second->required())?"Required":"Optional" );
      os <<". By default: "<<p->second->defValue() << '\n';
    } // for p

    os << "\n@param_file \t defines a file where the parameters are stored\n";
    os << '\n';

}

bool eoParser::userNeedsHelp(void)
{
  /*
     check whether there are long or short names entered
     without a corresponding parameter
  */
  // first, check if we want to check that !
  if (stopOnUnknownParam.value())
    {
      for (LongNameMapType::const_iterator lIt = longNameMap.begin(); lIt != longNameMap.end(); ++lIt)
	{
	  string entry = lIt->first;

	  MultiMapType::const_iterator it;

	  for (it = params.begin(); it != params.end(); ++it)
	    {
	      if (entry == it->second->longName())
		{
		  break;
		}
	    }

	  if (it == params.end())
	    {
	      string msg = "Unknown parameter: --" + entry + " entered, type -h or --help to see available parameters";
	      messages.push_back(msg);
	    }
	}

      for (ShortNameMapType::const_iterator sIt = shortNameMap.begin(); sIt != shortNameMap.end(); ++sIt)
	{
	  char entry = sIt->first;

	  MultiMapType::const_iterator it;

	  for (it = params.begin(); it != params.end(); ++it)
	    {
	      if (entry == it->second->shortName())
		{
		  break;
		}
	    }

	  if (it == params.end())
	    {
	      string entryString(1, entry);
	      string msg = "Unknown parameter: -" + entryString + " entered, type -h or --help to see available parameters";
	      messages.push_back(msg);
	    }
	}
    }
  return needHelp.value() || !messages.empty();
}

///////////////// I put these here at the moment
ostream & operator<<(ostream & _os, const eoParamParamType & _rate)
{
  _rate.printOn(_os);
  return _os;
}

istream & operator>>(istream & _is,  eoParamParamType & _rate)
{
  _rate.readFrom(_is);
  return _is;
}

